### PR TITLE
Pprint multiline formatting tweaks

### PR DESCRIPTION
**Pull request by danieldjohnson**
- Makes it so `DataCon` and `TypeCon` applications can have arguments on different lines
- Collapses together nested lambdas and `for` Hofs
- Aligns the arrows of nested pi types
- Adds/fixes multi-line pretty-printing for record, variant, tuple, and `Either` types

### CODE REVIEW

**Comment by apaszke** [comment]
Oh my god this is _exactly_ what I needed. Thanks!

File: src/lib/PPrint.hs [hunk 1]
```diff
@@ -294,6 +335,12 @@ instance PrettyPrec Atom where
     RecordTy items -> prettyExtLabeledItems items (line <> "&") ":"
     VariantTy items -> prettyExtLabeledItems items (line <> "|") ":"
 
+unInline :: Text -> Maybe Text
+unInline t = do
```
Comment (apaszke) [inline src/lib/PPrint.hs:339]: Why is it called `unInline` when it removes the parens?

**Review by apaszke — APPROVED** [review]

File: src/lib/PPrint.hs (hunk 1 above)
Comment (hexahedria) [reply src/lib/PPrint.hs:339]: Oh, that's a typo! I think I meant `unInfix`. I'll change it to that (unless you had a different suggestion?)

**Comment by hexahedria** [comment]
It seems like it makes a difference to me? Before:

```
>   ( takePicture @> ( ((m:Type) ?-> (Int32 & (Int32 & Bool)) -> (Scene m) -> (n:Int32) -> (((Fin 3) => Float32) & (Float32 & Float32)) -> (Fin n) => (Fin n) => (Fin 3) => Float32)
>   , (\m:Type.
>        \params:(Int32 & (Int32 & Bool)).
>          \scene:(Scene m).
>            \n:Int32.
>              \camera:(((Fin 3) => Float32) & (Float32 & Float32)).
>                tmp:Int32 = %fst params
>                tmp1:(Int32 & Bool) = %snd params
>                tmp2:Int32 = %fst tmp1
>                tmp3:Bool = %snd tmp1
>                tmp4:((((Fin 3) => Float32) & (Float32 & Float32)) -> (Fin n) => (Fin n) => Int64 -> (((Fin 3) => Float32) & ((Fin 3) => Float32))) =
>                  cameraRays n
>                tmp5:((Fin n) => (Fin n) => Int64 -> (((Fin 3) => Float32) & ((Fin 3) => Float32))) =
>                  tmp4 camera
>                tmp6:Int64 = newKey 0
...
```

After:
```
>   ( takePicture @> ( ((m:Type)
>                       ?-> (Int32 & (Int32 & Bool))
>                       -> (Scene m)
>                       -> (n:Int32)
>                       -> (((Fin 3) => Float32) & (Float32 & Float32))
>                       -> (Fin n) => (Fin n) => (Fin 3) => Float32)
>   , (\m:Type.
>        \params:(Int32 & (Int32 & Bool))
>             scene:(Scene m)
>             n:Int32
>             camera:(((Fin 3) => Float32) & (Float32 & Float32)).
>          tmp:Int32 = %fst params
>          tmp1:(Int32 & Bool) = %snd params
>          tmp2:Int32 = %fst tmp1
>          tmp3:Bool = %snd tmp1
>          tmp4:((((Fin 3) => Float32) & (Float32 & Float32))
>                -> (Fin n)
>                => (Fin n)
>                => Int64 -> (((Fin 3) => Float32) & ((Fin 3) => Float32))) =
>            cameraRays n
>          tmp5:((Fin n)
>                => (Fin n)
>                => Int64 -> (((Fin 3) => Float32) & ((Fin 3) => Float32))) =
>            tmp4 camera
>          tmp6:Int64 = newKey 0
...
```
Do you have ideas for things we could improve further here?

**Comment by hexahedria** [comment]
A contrived example where this makes a big difference:

```
data Foo a:Type = MkFoo a a a

def tripleFoo (x:a) : Foo a = MkFoo x x x

def doublePair (x:a) : (a & a) = (x, x)

%passes simp
:p tripleFoo $ doublePair $ tripleFoo $ doublePair $ tripleFoo 3
```

before:
```
>   evaluated bindings:
>   ( _ans_ @> ( (Foo ((Foo ((Foo Int32) & (Foo Int32))) & (Foo ((Foo Int32) & (Foo Int32)))))
>   , (MkFoo ((MkFoo ((MkFoo 3 3 3), (MkFoo 3 3 3)) ((MkFoo 3 3 3), (MkFoo 3 3 3)) ((MkFoo 3 3 3), (MkFoo 3 3 3))), (MkFoo ((MkFoo 3 3 3), (MkFoo 3 3 3)) ((MkFoo 3 3 3), (MkFoo 3 3 3)) ((MkFoo 3 3 3), (MkFoo 3 3 3)))) ((MkFoo ((MkFoo 3 3 3), (MkFoo 3 3 3)) ((MkFoo 3 3 3), (MkFoo 3 3 3)) ((MkFoo 3 3 3), (MkFoo 3 3 3))), (MkFoo ((MkFoo 3 3 3), (MkFoo 3 3 3)) ((MkFoo 3 3 3), (MkFoo 3 3 3)) ((MkFoo 3 3 3), (MkFoo 3 3 3)))) ((MkFoo ((MkFoo 3 3 3), (MkFoo 3 3 3)) ((MkFoo 3 3 3), (MkFoo 3 3 3)) ((MkFoo 3 3 3), (MkFoo 3 3 3))), (MkFoo ((MkFoo 3 3 3), (MkFoo 3 3 3)) ((MkFoo 3 3 3), (MkFoo 3 3 3)) ((MkFoo 3 3 3), (MkFoo 3 3 3))))) ) )
```

after:
```
>   evaluated bindings:
>   ( _ans_ @> ( (Foo
>                   ( (Foo ((Foo Int32) & (Foo Int32)))
>                   & (Foo ((Foo Int32) & (Foo Int32)))))
>   , (MkFoo
>        ( (MkFoo
>             ((MkFoo 3 3 3), (MkFoo 3 3 3))
>             ((MkFoo 3 3 3), (MkFoo 3 3 3))
>             ((MkFoo 3 3 3), (MkFoo 3 3 3)))
>        , (MkFoo
>             ((MkFoo 3 3 3), (MkFoo 3 3 3))
>             ((MkFoo 3 3 3), (MkFoo 3 3 3))
>             ((MkFoo 3 3 3), (MkFoo 3 3 3))))
>        ( (MkFoo
>             ((MkFoo 3 3 3), (MkFoo 3 3 3))
>             ((MkFoo 3 3 3), (MkFoo 3 3 3))
>             ((MkFoo 3 3 3), (MkFoo 3 3 3)))
>        , (MkFoo
>             ((MkFoo 3 3 3), (MkFoo 3 3 3))
>             ((MkFoo 3 3 3), (MkFoo 3 3 3))
>             ((MkFoo 3 3 3), (MkFoo 3 3 3))))
>        ( (MkFoo
>             ((MkFoo 3 3 3), (MkFoo 3 3 3))
>             ((MkFoo 3 3 3), (MkFoo 3 3 3))
>             ((MkFoo 3 3 3), (MkFoo 3 3 3)))
>        , (MkFoo
>             ((MkFoo 3 3 3), (MkFoo 3 3 3))
>             ((MkFoo 3 3 3), (MkFoo 3 3 3))
>             ((MkFoo 3 3 3), (MkFoo 3 3 3))))) ) )
```

File: src/lib/PPrint.hs (hunk 1 above)
Comment (hexahedria) [reply src/lib/PPrint.hs]: Changed to `fromInfix`

**Comment by apaszke** [comment]
This looks good so, I'm going to go ahead and merge it. The ray tracer issues are still in there I think --- if you scroll down into the computation you'll notice how deeply indented it gets --- but this is still a nice improvement. Thanks!

**Pull request merged**